pkg/storegateway/bucket.go

```diff
@@ -615,102 +620,329 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
+		// streaming chunks, that includes the series and chunks fetch duration for sending the streaming series.
+		stats.streamingSeriesEncodeResponseDuration += encodeDuration
+		stats.streamingSeriesSendResponseDuration += sendDuration
+		stats.streamingSeriesOtherDuration += time.Duration(util_math.Max(0, int64(time.Since(iterationBegin)-
+			stats.streamingSeriesFetchSeriesAndChunksDuration-encodeDuration-sendDuration)))
+	})
 
-				chunksCount += len(series.Chunks)
-				s.metrics.chunkSizeBytes.Observe(float64(chunksSize(series.Chunks)))
-			}
-			series.Labels = mimirpb.FromLabelsToLabelAdapters(lset)
-
-			// Encode the message. We encode it ourselves into a PreparedMsg in order to measure
-			// the time it takes.
-			encodeBegin := time.Now()
-			msg := &grpc.PreparedMsg{}
-			if err = msg.Encode(srv, storepb.NewSeriesResponse(&series)); err != nil {
-				err = status.Error(codes.Internal, errors.Wrap(err, "encode series response").Error())
-				return
+	seriesBuffer := make([]*storepb.StreamingSeries, req.StreamingChunksBatchSize)
+	for i := range seriesBuffer {
+		seriesBuffer[i] = &storepb.StreamingSeries{}
+	}
+	seriesBatch := &storepb.StreamingSeriesBatch{
+		Series: seriesBuffer[:0],
+	}
+	// TODO: can we send this in parallel while we start fetching the chunks below?
+	for seriesSet.Next() {
+		numSeries++
+		var lset labels.Labels
+		// Although subsequent call to seriesSet.Next() may release the memory of this series object,
+		// it is safe to hold onto the labels because they are not released.
+		lset, _ = seriesSet.At()
+
+		// We are re-using the slice for every batch this way.
+		seriesBatch.Series = seriesBatch.Series[:len(seriesBatch.Series)+1]
+		seriesBatch.Series[len(seriesBatch.Series)-1].Labels = mimirpb.FromLabelsToLabelAdapters(lset)
+
+		if len(seriesBatch.Series) == int(req.StreamingChunksBatchSize) {
+			err := s.sendMessage("streaming series", srv, storepb.NewStreamingSeriesResponse(seriesBatch), &encodeDuration, &sendDuration)
+			if err != nil {
+				return 0, err
 			}
-			encodeDuration += time.Since(encodeBegin)
+			seriesBatch.Series = seriesBatch.Series[:0]
+		}
+	}
+	if seriesSet.Err() != nil {
+		return 0, errors.Wrap(seriesSet.Err(), "expand series set")
+	}
+
+	// We need to send hints and stats before sending the chunks.
+	// Also, these need to be sent before we send IsEndOfSeriesStream=true.
+	if err := s.sendHintsAndStats(srv, resHints, stats); err != nil {
+		return 0, err
+	}
+
+	// Send any remaining series and signal that there are no more series.
+	seriesBatch.IsEndOfSeriesStream = true
+	err = s.sendMessage("streaming series", srv, storepb.NewStreamingSeriesResponse(seriesBatch), &encodeDuration, &sendDuration)
+	return numSeries, err
+}
+
+func (s *BucketStore) sendStreamingChunks(
+	req *storepb.SeriesRequest,
+	srv storepb.Store_SeriesServer,
+	it seriesChunksSetIterator,
+	stats *safeQueryStats,
+) (seriesCount, chunksCount int, err error) {
```

Comment:
Can we use the fields within `stats` instead of returning `seriesCount, chunksCount int`?